### ISSUE TITLE
Gateway is stripping operation names when forwarding to federated service

### ISSUE THREAD

**Issue by rtymchyk** [apollographql/apollo-server#4169]
I'm not quite sure if this is a bug or a question, but I can observe that if apollo-client fires a query like this:
```
{
  "operationName": "AuthenticatedRouteViewerQuery",
  "variables": {},
  "query": "query AuthenticatedRouteViewerQuery { viewer { user { __typename id } } }"
}
```

And it passes through the gateway, the back-end ruby/rails service reports:
```
Processing by GraphqlController#execute as */*
Parameters: {"query"=>"{viewer{user{id __typename}__typename}}", "variables"=>{}, "graphql"=>{"query"=>"{viewer{user{id __typename}__typename}}", "variables"=>{}}}
```

I am running:
`@apollo/gateway@0.14.1`
`apollo-server@^2.12.0`

Clearly operation names are intact because I can see them in the Apollo Graph Manager metrics section, but downstream they are not visible in the actual service.

**Comment by zebulonj** [comment]
For what it's worth, this question came up on my team and this is a behavior I would expect. Since the gateway fragments the original query, I wouldn't have any expectations about what operation name (if any) the implementing services would observe on those fragments.

**Comment by andrewmcgivery** [comment]
Agreed with @zebulonj. We discussed this internally in our team and we also concluded that this is probably expected because the queries that get sent to downstream services are completely different than the original query due to how Gateway sends queries, query batching, and fragmentation. Bit of a trade off either way.

**Comment by Kartikkumargujarati** [comment]
Hey @rtymchyk, did you ever figure out a way to forward the operation name from gateway to a federated service?

**Comment by rtymchyk** [comment]
> Hey @rtymchyk, did you ever figure out a way to forward the operation name from gateway to a federated service?

I haven't.

**Comment by Kartikkumargujarati** [comment]
I was looking into the `apollo-server` code and see that `[operation](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-core/src/requestPipeline.ts#L333)` returned by `[getOperationAST](https://github.com/graphql/graphql-js/blob/master/src/utilities/getOperationAST.js#L9)` function is used to set the `operationName` on `requestContext` object. 

Based on this, I was able to pull the operation name from the document definition by doing `requestContext.document.definitions[0].selectionSet.selections[0].name.value`. Used https://astexplorer.net/#
I am not sure if this is a recommended approach though.

@abernix @andrewmcgivery @zebulonj Do you have any other suggestions or a cleaner way to pull the operation name from the requestContext object? 

My use case is, I am trying to do report a metric when `willSendResponse` is called on the federated service and want to include the operation name on that metric.

**Comment by rtymchyk** [comment]
@Kartikkumargujarati Did you end up weeding out the operation name yourself?

**Comment by Kartikkumargujarati** [comment]
@rtymchyk Given I haven't found a better way than what I mentioned above, and haven't heard back from maintainers, I might as well just weed out the operation myself. I will post back here if I end up finding anything cleaner.

**Comment by Kartikkumargujarati** [comment]
Folks, I found a more "desirable" workaround for this problem.
In your gateway:
`
class CustomDataSource extends RemoteGraphQLDataSource {
  willSendRequest({ request, context }) {
    // Get the operation name from the context
    const operationName = context.req.body.operationName;
    // Add the above operation name as http header to the request
    request.http.headers.set('operationName', operationName);
  }
}
`

In your federated service:
`
const server = new ApolloServer({ 
context: ({ req }) => {
    // Get the operation name from the headers
    const operationName = req.header("operationName") || '';
    return { operationName }; // or do whatever you wish to do
  }
typeDefs, 
resolvers
});
`
Hope this helps you.

**Comment by Kartikkumargujarati** [comment]
Folks, I found a more "desirable" workaround for this problem.
In your gateway:
```
class CustomDataSource extends RemoteGraphQLDataSource {
  willSendRequest({ request, context }) {
    // Get the operation name from the context
    const operationName = context.req.body.operationName;
    // Add the above operation name as http header to the request
    request.http.headers.set('operationName', operationName);
  }
}
```

In your federated service:
```
const server = new ApolloServer({ 
context: ({ req }) => {
    // Get the operation name from the headers
    const operationName = req.header("operationName") || '';
    return { operationName }; // or do whatever you wish to do
  }
typeDefs, 
resolvers
});
```
Hope this helps you.

**Comment by Kartikkumargujarati** [comment]
Folks, I found a more "desirable" workaround for this problem. Essentially, passing the operation name as a header in the request.

In your gateway:

class CustomDataSource extends RemoteGraphQLDataSource {
  willSendRequest({ request, context }) {
    // Get the operation name from the context
    const operationName = context.req.body.operationName;
    // Add the above operation name as http header to the request
    request.http.headers.set('operationName', operationName);
  }
}

In your federated service:

const server = new ApolloServer({ 
context: ({ req }) => {
    // Get the operation name from the headers
    const operationName = req.header("operationName") || '';
    return { operationName }; // or do whatever you wish to do
  }
typeDefs, 
resolvers
});

Hope this helps you.

**Comment by Kartikkumargujarati** [comment]
Folks, I found a more "desirable" workaround for this problem. Essentially, passing the operation name as a header in the request.

In your gateway:
```
class CustomDataSource extends RemoteGraphQLDataSource {
  willSendRequest({ request, context }) {
    // Get the operation name from the context
    const operationName = context.req.body.operationName;
    // Add the above operation name as http header to the request
    request.http.headers.set('operationName', operationName);
  }
}
```

In your federated service:
```
const server = new ApolloServer({ 
context: ({ req }) => {
    // Get the operation name from the headers
    const operationName = req.header("operationName") || '';
    return { operationName }; // or do whatever you wish to do
  }
typeDefs, 
resolvers
});
```

Hope this helps you.

**Comment by rtymchyk** [comment]
> Folks, I found a more "desirable" workaround for this problem. Essentially, passing the operation name as a header in the request.
> 
> In your gateway:
> ```
> class CustomDataSource extends RemoteGraphQLDataSource {
>   willSendRequest({ request, context }) {
>     // Get the operation name from the context
>     const operationName = context.req.body.operationName;
>     // Add the above operation name as http header to the request
>     request.http.headers.set('operationName', operationName);
>   }
> }
> ```
> 
> In your federated service:
> ```
> const server = new ApolloServer({ 
> context: ({ req }) => {
>     // Get the operation name from the headers
>     const operationName = req.header("operationName") || '';
>     return { operationName }; // or do whatever you wish to do
>   }
> typeDefs, 
> resolvers
> });
> ```
> 
> Hope this helps you.

Good idea! Thanks :)
